SPARK-5018 [MLlib] [WIP] Make MultivariateGaussian public

tgaloppo opened the pull request:
Moving MutlivariateGaussian from private[mllib] to public.  The class uses Breeze vectors internally, so this involves creating a public interface using MLlib vectors and matrices.

This initial commit provides public construction, accessors for mean/covariance, density and log-density.

Other potential methods include entropy and sample generation.

tgaloppo committed:
SPARK-5018

MultivariateGaussian.scala - Made class public and exposed public methods leveraging MLlib vectors and matrices. Added logpdf method providing log-density calculation.

MultivariateGaussianSuite.scala - Test are now performed through the public methods.

tgaloppo committed:
SPARK-5018

SparkQA commented:
[Test build #25135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25135/consoleFull) for   PR 3923 at commit [`0943dc4`](https://github.com/apache/spark/commit/0943dc43b887a2626d6e6ef6cf0f8ac48ebc7f0f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25135/consoleFull) for   PR 3923 at commit [`0943dc4`](https://github.com/apache/spark/commit/0943dc43b887a2626d6e6ef6cf0f8ac48ebc7f0f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25135/
Test PASSed.

jkbradley commented:
@tgaloppo Thanks for the PR!  I should have said in the JIRA: We currently do not expose Breeze publicly (so that we are not locked into a particular linalg library).  For this API, we'll need to expose only mllib Vector and Matrix types (though we can use Breeze internally for learning).

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,22 +31,57 @@ import org.apache.spark.mllib.util.MLUtils
+class MultivariateGaussian private[mllib] (
     val mu: DBV[Double], 
     val sigma: DBM[Double]) extends Serializable {
 
```
I'd add a check here to make sure that the dimensions of mu, sigma match up.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,22 +31,57 @@ import org.apache.spark.mllib.util.MLUtils
+   * @param mu The mean vector of the distribution
+   * @param sigma The covariance matrix of the distribution
+   */
+  def this(mu: Vector, sigma: Matrix) = {
```
Maybe mu, sigma should be named mean, covariance here to match the getter methods.  (Or those could be renamed; either is OK with me.)

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,22 +31,57 @@ import org.apache.spark.mllib.util.MLUtils
    */
   private val (rootSigmaInv: DBM[Double], u: Double) = calculateCovarianceConstants
   
+  /** Return the mean vector for this distribution */
```
I'd include "mu" in the doc since it's such a common variable name.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,22 +31,57 @@ import org.apache.spark.mllib.util.MLUtils
   private val (rootSigmaInv: DBM[Double], u: Double) = calculateCovarianceConstants
   
+  /** Return the mean vector for this distribution */
+  def getMean(): Vector = {
```
Exclude parentheses () since it's just an accessor method

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,22 +31,57 @@ import org.apache.spark.mllib.util.MLUtils
+    Vectors.fromBreeze(mu)
+  }
+  
+  /** Return the covariance matrix for this distribution */
```
ditto: "sigma" in doc

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,22 +31,57 @@ import org.apache.spark.mllib.util.MLUtils
+  }
+  
+  /** Return the covariance matrix for this distribution */
+  def getCovariance(): Matrix = {
```
Exclude parentheses () since it's just an accessor method

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -54,7 +90,7 @@ private[mllib] class MultivariateGaussian(
    * 
    * We here compute distribution-fixed parts 
-   *  (2*pi)^(-k/2) * det(sigma)^(-1/2)
+   *  log((2*pi)^(-k/2) * det(sigma)^(-1/2))
```
By the way, in Scala doc (here and elsewhere), I believe the correct way to write exponents is to have an opening ^ and closing ^ around the exponent:
```
det(sigma)^(-1/2)^
```

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -91,7 +127,7 @@ private[mllib] class MultivariateGaussian(
       val pinvS = diag(new DBV(d.map(v => if (v > tol) math.sqrt(1.0 / v) else 0.0).toArray))
     
-      (pinvS * u, math.pow(2.0 * math.Pi, -mu.length / 2.0) * math.pow(pdetSigma, -0.5))
+      (pinvS * u, math.log(math.pow(2.0 * math.Pi, -mu.length / 2.0) * math.pow(pdetSigma, -0.5)))
```
This should skip the log(pow("stuff")) combination and just have "stuff" (with the log moved inside of "stuff"), in order to skip the math.pow call.
Also, since you'll need log(pdetSigma), you can compute that from the beginning, using a sum instead of multiplying in line 124.

tgaloppo replied:
Calculating log(pdetSigma) that way will incur a performance penalty (calling log for each singular value), but it is probably more numerically stable.

jkbradley replied:
True, it will be slower, but I assume eigSym will dominate the running time.

jkbradley commented:
@tgaloppo It's looking good.  Mostly small comments, except for the first 2.  Your call about naming mu,sigma vs. mean,covariance.

tgaloppo committed:
Fixed accessor methods to match member variable names.
Modified calculations to avoid log(pow(x,y)) calculations

SparkQA commented:
[Test build #25187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25187/consoleFull) for   PR 3923 at commit [`8c35381`](https://github.com/apache/spark/commit/8c353810670690b3814b298523ecfbcfacc5b0f0).
 * This patch merges cleanly.

tgaloppo committed:
Rearranged equation for part of density function

SparkQA commented:
[Test build #25190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25190/consoleFull) for   PR 3923 at commit [`91a5fae`](https://github.com/apache/spark/commit/91a5fae036f668e79422d6003ba62fc973d003a8).
 * This patch merges cleanly.

tgaloppo commented:
@jkbradley Thanks! I have made the requested changes.  Are there any other public methods that you think would be useful to add at this time?

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,33 +31,72 @@ import org.apache.spark.mllib.util.MLUtils
     val mu: DBV[Double], 
     val sigma: DBM[Double]) extends Serializable {
 
+  if (sigma.cols != sigma.rows) {
```
This is fine, but you can also use ```require``` for checking input arguments (since that throws ```IllegalArgumentException``` and is a bit shorter).

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -84,14 +124,14 @@ private[mllib] class MultivariateGaussian(
-      // pseudo-determinant is product of all non-zero singular values
-      val pdetSigma = d.activeValuesIterator.filter(_ > tol).reduce(_ * _)
+      // log(pseudo-determinant) is sum of the logs of all non-zero singular values
+      val logPseudoDetSigma = d.activeValuesIterator.filter(_ > tol).map(math.log(_)).reduce(_ + _)
```
```map(math.log(_))``` can be simplified to ```map(math.log)```
Also, ```reduce(_ + _)``` can be replaced with ```sum```

jkbradley commented:
@tgaloppo  Two small comments, but they are just superficial.  After those, I think it will be ready.  I don't see a need to add other methods for now.  Thanks!

tgaloppo committed:
Style improvements

tgaloppo commented:
@jkbradley Changes made.  Thanks!

SparkQA commented:
[Test build #25261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25261/consoleFull) for   PR 3923 at commit [`9fa3bb7`](https://github.com/apache/spark/commit/9fa3bb74b0834449d9ccb5ecf54b95d1ecc33db2).
 * This patch merges cleanly.

jkbradley commented:
@tgaloppo   Thanks!

LGTM pending Jenkins tests

CC: @mengxr

SparkQA commented:
[Test build #25261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25261/consoleFull) for   PR 3923 at commit [`9fa3bb7`](https://github.com/apache/spark/commit/9fa3bb74b0834449d9ccb5ecf54b95d1ecc33db2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25261/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.mllib.stat.impl
 
 import breeze.linalg.{DenseVector => DBV, DenseMatrix => DBM, diag, max, eigSym}
 
+import org.apache.spark.mllib.linalg.{ Vectors, Vector, Matrices, Matrix }
```
Remove space after `{` and before `}`.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,33 +31,67 @@ import org.apache.spark.mllib.util.MLUtils
  * @param sigma The covariance matrix of the distribution
  */
-private[mllib] class MultivariateGaussian(
+class MultivariateGaussian private[mllib] (
```
If we want to make it public, should it still under `stat.impl`? Maybe is it good to put it under `stat.distribution` and mark it as `@DeveloperApi` since we don't have a `Distribution` trait yet.

jkbradley replied:
+1

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/stat/impl/MultivariateGaussian.scala:
```diff
@@ -30,33 +31,67 @@ import org.apache.spark.mllib.util.MLUtils
  */
-private[mllib] class MultivariateGaussian(
+class MultivariateGaussian private[mllib] (
     val mu: DBV[Double], 
```
Those are public fields of the class. We should not expose breeze types.

tgaloppo committed:
Made mu, sigma private[mllib] members of MultivariateGaussian
Moved MultivariateGaussian (and test suite) from stat.impl to stat.distribution (required updates in GaussianMixture{EM,Model}.scala)
Marked MultivariateGaussian as @DeveloperApi
Fixed style error

SparkQA commented:
[Test build #25339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25339/consoleFull) for   PR 3923 at commit [`e30a100`](https://github.com/apache/spark/commit/e30a100c2e5f3fd6af7f7e83b48315c44437147f).
 * This patch **does not merge cleanly**.

SparkQA commented:
[Test build #25339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25339/consoleFull) for   PR 3923 at commit [`e30a100`](https://github.com/apache/spark/commit/e30a100c2e5f3fd6af7f7e83b48315c44437147f).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25339/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/stat/distribution/MultivariateGaussian.scala:
```diff
@@ -15,11 +15,13 @@
 
+import org.apache.spark.mllib.linalg.{Vectors, Vector, Matrices, Matrix}
 import org.apache.spark.mllib.util.MLUtils
+import org.apache.spark.annotation.DeveloperApi;
```
sort import alphabetically

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/stat/distribution/MultivariateGaussian.scala:
```diff
@@ -15,11 +15,13 @@
+import org.apache.spark.annotation.DeveloperApi;
 
 /**
  * This class provides basic functionality for a Multivariate Gaussian (Normal) Distribution. In
```
Please add `:: DeveloperApi ::` before `This class ...`. We need it to generate the doc correctly. You can check other `@DeveloperApi` usage as examples.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/stat/distribution/MultivariateGaussian.scala:
```diff
@@ -30,33 +32,68 @@ import org.apache.spark.mllib.util.MLUtils
-    val sigma: DBM[Double]) extends Serializable {
+@DeveloperApi
+class MultivariateGaussian private[mllib] (
+    private[mllib] val mu: DBV[Double], 
```
Instead of having `mu`/`sigma` private and add getters, could we make them MLlib vector/matrix types and add private members of breeze types? Then we can make this constructor public and remove getters. The overhead is little because we don't copy the data arrays.

mengxr commented:
@tgaloppo Besides inline comments, please resolve conflicts with the master branch. The patch does not merge cleanly.

tgaloppo committed:
Merge remote-tracking branch 'upstream/master' into spark-5018

Conflicts:
	mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixtureEM.scala

SparkQA commented:
[Test build #25360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25360/consoleFull) for   PR 3923 at commit [`b4121b4`](https://github.com/apache/spark/commit/b4121b4e43c8ae5ff507301b3818be9a557548a1).
 * This patch merges cleanly.

tgaloppo commented:
I have made the requested changes and resolved the merge conflicts.

Question: MutlivariateGuassian now keeps a private Breeze version of the mean vector rather than convert the MLlib version to a Breeze vector with each call to {log}pdf().  Is this worthwhile? Or is the .toBreeze.toDenseVector sequence negligible enough that this is unnecessary?

CC: @mengxr @jkbradley

SparkQA commented:
[Test build #25360 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25360/consoleFull) for   PR 3923 at commit [`b4121b4`](https://github.com/apache/spark/commit/b4121b4e43c8ae5ff507301b3818be9a557548a1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25360/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixtureEM.scala:
```diff
@@ -17,15 +17,16 @@
+import org.apache.spark.rdd.RDD
 import org.apache.spark.util.Utils
 
+import scala.collection.mutable.IndexedSeq
```
Organize imports (scala/java come first)

jkbradley commented:
Thanks for the updates!  LGTM except for organizing the imports in my 1 comment.

As far as I understand, the conversions to and from Breeze should be efficient since they mainly involve copying references, not the underlying data (as long as the representations remain the same: dense to dense, or sparse to sparse).

tgaloppo committed:
Style correction

SparkQA commented:
[Test build #25367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25367/consoleFull) for   PR 3923 at commit [`2b15587`](https://github.com/apache/spark/commit/2b1558727af2a28c7c3aa675fbdb040cd17fee01).
 * This patch merges cleanly.

tgaloppo commented:
Thanks, @jkbradley 
I have made the style correction.

SparkQA commented:
[Test build #25367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25367/consoleFull) for   PR 3923 at commit [`2b15587`](https://github.com/apache/spark/commit/2b1558727af2a28c7c3aa675fbdb040cd17fee01).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25367/
Test PASSed.

mengxr commented:
Merged into master. Thanks!
